### PR TITLE
Add statistics to dashboard for all pipelines

### PR DESCRIPTION
**Pull request by magickris93**

### CODE REVIEW

File: dashboard/templates/dashboard.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        <thead class="thead-dark">
+        <tr>
+          <th scope="col">Pipeline</th>
+          <th scope="col">Buildsets</th>
+        </tr>
+      </thead>
+      <tbody>
+        {% for pipe in pipeline_stats %}
+            <tr>
+                <td><a class="mt-5 font-weight-bold" href="{{ url_for('status.show_status', pipename=pipe)}}">{{ pipe }}</a></td>
```
Comment (kam193) [inline dashboard/templates/dashboard.html]: Add one space before ``}}`` with url

File: dashboard/service.py [hunk 2]
```diff
@@ -17,9 +17,9 @@ def fetch_json_data(endpoint):
     return res.json()
 
 
-def parse_overall_stats(resource):
+def pipelines_list(resource):
```
Comment (kam193) [inline dashboard/service.py]: I think, this name is a bit confusing. 1) this function returns dict, not list; 2) it's dict of stats for every pipeline. For me, "pipelines_list" doesn't match to this.

File: dashboard/service.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    res = requests.get(endpoint, timeout=3)
+
+    if res.status_code not in [200, 304]:
+        current_app.logger.error(res.text)
+        raise RemoteServerError('Request for Zuul status failed.')
+
+    return res.json()
+
+
+def pipelines_list(resource):
```
Comment (kawazaki0) [inline dashboard/service.py]: Maybe `pipelines_stats`?

File: dashboard/controller.py [hunk 4]
```diff
@@ -34,21 +31,22 @@ def error_404(error):
...
-        current_app.logger.error(res.text)
-        raise RemoteServerError('Request for Zuul status failed.')
 
-    res = res.json()
-    queues = make_queues(res['pipelines'], pipename)
-    return render_template('status.html', queues=queues, pipename=pipename)
+@status.route('/', methods=['GET'])
+def show_dashboard():
+    url = str(f'{config["zuul"]["url"].rstrip("/")}/'
+              f'{config["zuul"]["status_endpoint"]}')
```
Comment (kawazaki0) [inline dashboard/controller.py]: extract as a function. And maybe add test for it

File: tests/test_service.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+import json
+import unittest
+
+from dashboard import service
+
+
+class TestService(unittest.TestCase):
```
Comment (kawazaki0) [inline tests/test_service.py]: You should add tests for other functions in service.py. Now, tests for controller.py are testing cases that should be here (e.g. raising exceptions when request for Zuul fails). In test_service.py should be unit tests for service.py, and test_controller.py could be integration tests for controller.py (they can be more targeted to testing that all modules are connected correctly, and less targeted to check every corner case.

File: tests/test_service.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+                    'test_pipeline2': {'buildsets_count': 0}}
+        path = str('tests/static/'
+                   'status_pipelines_with_no_buildsets.json')
+        with open(path, 'r') as f:
+            resource = json.load(f)
+        result = service.pipelines_list(resource=resource['pipelines'])
+        self.assertDictEqual(result, expected)
+
+    def test_pipelines_list_returns_expected_for_queue_with_many_heads(self):
+        expected = {'test_pipeline1': {'buildsets_count': 3}}
```
Comment (kawazaki0) [inline tests/test_service.py]: I like the given-when-then structure in tests whenever it could be applied. In this pattern, the `expected` should be in the last section.

File: dashboard/controller.py (hunk 4 above)
Comment (pawelzny) [reply dashboard/controller.py]: Agree. This should be moved to service as for example `status_endpoint()` function.

File: dashboard/service.py (hunk 3 above)
Comment (pawelzny) [inline dashboard/service.py]: Argument of this function points to "resources" but now it gets only "pipelines", the same as in `make_queues` function.
IMHO in this case argument should be renamed to `pipelines`.

File: dashboard/service.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    pipeline_stats = {}
```
Comment (pawelzny) [inline dashboard/service.py]: I think a list of named tuples will make more sense instead of dictionary.
I would like to see list of Pipeline named tuple which holds name and buildsets counter.
At the very end in template you are iterating over dictionary which is not optimal. List will be better.

File: dashboard/templates/base.html [hunk 8]
```diff
@@ -17,13 +17,9 @@
                 aria-expanded="false" aria-label="Toggle navigation">
             <span class="navbar-toggler-icon"></span>
         </button>
-
         <div class="collapse navbar-collapse" id="navbarSupportedContent">
             <ul class="navbar-nav mr-auto">
                 <li class="nav-item active">
-                    <a class="nav-link" href="{{ url_for('status.show_status') }}">Status</a>
```
Comment (pawelzny) [inline dashboard/templates/base.html]: We need to rollback this change and make it as it was before.

File: dashboard/templates/dashboard.html [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    <div class="lead mx-auto w-25 mt-5 table-responsive">
+      <table class="table">
+        <thead class="thead-dark">
+        <tr>
+          <th scope="col">Pipeline</th>
+          <th scope="col">Buildsets</th>
+        </tr>
+      </thead>
+      <tbody>
+        {% for pipe in pipeline_stats %}
```
Comment (pawelzny) [inline dashboard/templates/dashboard.html]: There is no pipes. Those are **pipelines**.
